SPECIFICATION.md

```diff
@@ -0,0 +1,154 @@
+(PCBs) in mind. PCBs are usually manufactured using the Gerber format as an
+exchange format to describe the etching layers, which does not support
+quadratic / cubic curves.
+
+The main inspiration for this font format was the LFF font used by LibreCAD.
+
+
+## High Level Format Description
+
+Every FontoBene font file consists of UTF-8 encoded text. The lines are
+separated by newlines (U+000A LINE FEED). It starts with a header, followed by
+glyph definitions.
+
+
+### Header
+
+The header is parsed line-by-line, starting with a number sign (`#`) as the
+first character on every line. The number sign is followed by a space, a key,
+and a value. Key and value are separated by a colon (`:`). Everything following
+the colon is considered the value. Both keys and values are strings.
+
+Example:
+
+    # Format: FontoBene/1.0
+    # Creator: Max Müller
+    # Version: 1.0
+    # VersionDate: 2017-09-27
+    # License: GPLv3+
+
+All keys that may be used are specified in the specification. A user MAY use
+non-standard keys, but must prefix them with `X-`, for example `X-Foo: Bar`.
+
+Whitespace around key and value should be stripped by the parsers.
+
+
+### Body
+
+The first empty line separates the header from the body. The body consists of
+glyph definition blocks separated by at least one empty line.
+
+Every glyph definition starts with the glyph declaration, a unicode codepoint
+in square brackets (e.g. `[1F4A9]` for the pile of poo character). It may
+optionally be followed by a UTF-8 representation of the character for easier
+previewing (e.g. `💩`). However, that preview will be ignored by the parser.
+
+The notation used inside the brackets follows the convention described in [*The
+Unicode Standard, Version 10.0, Appendix A, Notational Conventions, Code
+Points*][unicode-10-spec]:
+
+> In running text, an individual Unicode code point is expressed as U+n, where
+> n is four to six hexadecimal digits, using the digits 0–9 and uppercase
+> letters A–F (for 10 through 15, respectively).
+>
+> Leading zeros are omitted, unless the code point would have fewer than four
+> hexadecimal digits—for example, U+0001, U+0012, U+0123, U+1234, U+12345,
+> U+102345.
+>
+> The U+ may be omitted for brevity in tables or when denoting ranges.
+
+The glyph declaration is followed by a list of polylines or glyph references,
+each on a single line.
+
+A polyline looks like this:
+
+    0,0;3,9;6,0
+
+It consists of coordinate pairs, separated by semicolons (`;`). The X and Y
+coordinates are separated with a comma (`,`).  The coordinates are arbitrary
+precision floating point numbers using the full stop (U+002E) as decimal mark.
+If a coordinate is a whole number, the decimal mark and the fractional part of
+the number may be omitted.
+
+The coordinate system starts at the lower left corner. Positive X coordinates
+point to the right. Positive Y coordinates point to the top. Negative
+coordinates are not allowed.
```

Comment:
Why are negative values not allowed, resp. where is the baseline of characters like "p", q" and so on?

Reply:
That's a good point, I'll still have to define that. There are three ways how we could handle alignment and scaling:

- We could use a reference height for characters, e.g. in the range 0 (from the descender) to 100 (at the ascender). The baseline would be e.g. at 20. That would save us from having to parse negative numbers.
- We could allow negative numbers and specify y=0 as the baseline and e.g. 120 as the ascender line. That would also make the height fixed.
- We could allow the negative numbers and specify y=0 as the baseline, without specifying the height of a character. The font would include a `scaling_factor` in the header though, which would be multiplied with the coordinates to achieve a standardized glyph height.

![img](http://www.loveofgraphics.com/wp-content/uploads/2011/10/AnatomyofLetter.png)

Reply:
At the moment I cannot say which variant I would prefer ;) 

But it seems that the coordinate system is the most complicated and essential part of our font specification. Maybe it would make sense to document the concept of the coordinate system (incl. images) in a separate markdown file too keep the file format specification more clean. Or what do you think @dbrgn?

Btw, in case of the font is used for PCBs or other fabrication related things, It's very important to have the text size known in a physical length unit (e.g. millimeters). I don't know much about fonts, so I'm not sure how the relation between point, pixel and physical size exactly is. In which unit should the coordinates in the font file be defined? And should the parser provide a way for unit conversion?

Reply:
> In which unit should the coordinates in the font file be defined?

In none :) I'd just specify a range, e.g. 0-100 for baseline to cap height. Then your software can scale the font as desired.

Having a fixed height range (0-100) requires more work when doing font conversions, but it's much easier to handle than having to deal with a scale factor in the metadata. I think 0-100 for base-cap height is my current favorite. The fonts could go a bit over / below these values if desired, but it's in their responsibility to keep the fonts readable.

Reply:
> I'd just specify a range, e.g. 0-100 for baseline to cap height. Then your software can scale the font as desired.

Oh, so simple and still powerful :see_no_evil: :smiley: 

But probably we still need to define some more things, because in the LFF header one can find these additional parameters:
```
# LetterSpacing:     3
# WordSpacing:       6.75
# LineSpacingFactor: 1
```

Reply:
I think at least the space character (" ") is a problem, because it's calculated width would be zero (no polylines) which is wrong. LFF seems to define the width of the space character with the attribute "LetterSpacing" in the header.

Reply:
Yeah, I wonder how we should deal with that. I think the space character is the only one that we need to consider. It could also be derived from the width of a specific glyph though, e.g. the capital A? Not sure if this is something that is being done by font layout engines or not.